File: server/exchanges/ExchangeManager.js
const BinanceConnector = require('./connectors/BinanceConnector');
const CoinbaseConnector = require('./connectors/CoinbaseConnector');
const KrakenConnector = require('./connectors/KrakenConnector');
const BitfinexConnector = require('./connectors/BitfinexConnector');
const BybitConnector = require('./connectors/BybitConnector');
const OKXConnector = require('./connectors/OKXConnector');

class ExchangeManager {
  constructor() {
    this.connectors = {
      binance: new BinanceConnector(),
      coinbase: new CoinbaseConnector(),
      kraken: new KrakenConnector(),
      bitfinex: new BitfinexConnector(),
      bybit: new BybitConnector(),
      okx: new OKXConnector()
    };
  }

  async subscribe(exchange, symbol, callback) {
    if (!this.connectors[exchange]) {
      throw new Error(`Exchange ${exchange} not supported`);
    }
    
    return await this.connectors[exchange].subscribe(symbol, callback);
  }

  unsubscribe(exchange, symbol) {
    if (this.connectors[exchange]) {
      this.connectors[exchange].unsubscribe(symbol);
    }
  }

  getAvailableExchanges() {
    return Object.keys(this.connectors).map(exchange => ({
      id: exchange,
      name: exchange.charAt(0).toUpperCase() + exchange.slice(1),
      supportedPairs: this.connectors[exchange].getSupportedPairs()
    }));
  }
}

module.exports = ExchangeManager;

File: python_server/exchanges/exchange_manager.py
"""Exchange Manager - Handles connections to multiple exchanges"""

import logging
from typing import Dict, Any, Callable, List

from .connectors.binance_connector import BinanceConnector
from .connectors.coinbase_connector import CoinbaseConnector
from .connectors.kraken_connector import KrakenConnector

logger = logging.getLogger(__name__)


class ExchangeManager:
    """Manages connections to multiple cryptocurrency exchanges"""

    def __init__(self):
        """Initialize exchange connectors"""
        self.connectors = {
            'binance': BinanceConnector(),
            'coinbase': CoinbaseConnector(),
            'kraken': KrakenConnector(),
        }
        logger.info(f"Initialized {len(self.connectors)} exchange connectors")

    def subscribe(self, exchange: str, symbol: str, callback: Callable) -> Dict[str, Any]:
        """Subscribe to orderbook updates for a symbol on an exchange"""
        if exchange not in self.connectors:
            raise ValueError(f"Exchange {exchange} not supported")

        logger.info(f"Subscribing to {exchange} {symbol}")
        return self.connectors[exchange].subscribe(symbol, callback)

    def unsubscribe(self, exchange: str, symbol: str):
        """Unsubscribe from orderbook updates"""
        if exchange in self.connectors:
            logger.info(f"Unsubscribing from {exchange} {symbol}")
            self.connectors[exchange].unsubscribe(symbol)

    def get_available_exchanges(self) -> List[Dict[str, Any]]:
        """Get list of available exchanges and their supported pairs"""
        exchanges = []
        for exchange_id, connector in self.connectors.items():
            exchanges.append({
                'id': exchange_id,
                'name': exchange_id.capitalize(),
                'supportedPairs': connector.get_supported_pairs()
            })
        return exchanges

    def close_all(self):
        """Close all exchange connections"""
        for connector in self.connectors.values():
            connector.close()
